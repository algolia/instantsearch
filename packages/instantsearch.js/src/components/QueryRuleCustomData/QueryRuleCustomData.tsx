/** @jsx h */

import { h } from 'preact';

import Template from '../Template/Template';

import type { ComponentCSSClasses } from '../../types';
import type {
  QueryRuleCustomDataCSSClasses,
  QueryRuleCustomDataTemplates,
} from '../../widgets/query-rule-custom-data/query-rule-custom-data';

export type QueryRuleCustomDataComponentCSSClasses =
  ComponentCSSClasses<QueryRuleCustomDataCSSClasses>;

export type QueryRuleCustomDataComponentTemplates =
  Required<QueryRuleCustomDataTemplates>;

export type QueryRuleCustomDataProps = {
  cssClasses: QueryRuleCustomDataComponentCSSClasses;
  templates: QueryRuleCustomDataComponentTemplates;
  items: any[];
};

const QueryRuleCustomData = ({
  cssClasses,
  templates,
  items,
}: QueryRuleCustomDataProps) => (
  <Template
    templateKey="default"
    templates={templates}
    rootProps={{ className: cssClasses.root }}
    data={{ items }}
  />
);

export default QueryRuleCustomData;
