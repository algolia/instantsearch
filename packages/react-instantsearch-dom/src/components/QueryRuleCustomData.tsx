import classNames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

import { createClassNames } from '../core/utils';

type CustomUserData = {
  [key: string]: any;
};

const cx = createClassNames('QueryRuleCustomData');

type QueryRuleCustomDataRenderProps<TItem> = {
  items: TItem[];
};

export type QueryRuleCustomDataProps<TItem> = {
  items: TItem[];
  className?: string;
  children: (options: QueryRuleCustomDataRenderProps<TItem>) => React.ReactNode;
};

const QueryRuleCustomData: React.FC<QueryRuleCustomDataProps<CustomUserData>> =
  ({ items, className, children }) => (
    <div className={classNames(cx(''), className)}>{children({ items })}</div>
  );

QueryRuleCustomData.propTypes = {
  items: PropTypes.arrayOf(PropTypes.object).isRequired as any,
  className: PropTypes.string,
  children: PropTypes.func.isRequired,
};

export default QueryRuleCustomData;
