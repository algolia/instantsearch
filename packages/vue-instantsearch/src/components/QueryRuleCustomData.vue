<template>
  <div v-if="state" :class="suit()">
    <slot :items="state.items">
      <div v-for="(item, key) in state.items" :key="key">
        <slot name="item" :item="item">
          <pre>{{ item }}</pre>
        </slot>
      </div>
    </slot>
  </div>
</template>

<script>
import { connectQueryRules } from 'instantsearch.js/es/connectors';

import { createSuitMixin } from '../mixins/suit';
import { createWidgetMixin } from '../mixins/widget';

export default {
  name: 'AisQueryRuleCustomData',
  mixins: [
    createSuitMixin({ name: 'QueryRuleCustomData' }),
    createWidgetMixin(
      {
        connector: connectQueryRules,
      },
      {
        $$widgetType: 'ais.queryRuleCustomData',
      }
    ),
  ],
  props: {
    transformItems: {
      type: Function,
      required: false,
      default: undefined,
    },
  },
  computed: {
    widgetParams() {
      return {
        transformItems: this.transformItems,
      };
    },
  },
};
</script>
