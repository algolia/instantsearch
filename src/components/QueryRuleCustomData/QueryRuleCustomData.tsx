import React from 'preact-compat';
import {
  QueryRuleCustomDataCSSClasses,
  QueryRuleCustomDataTemplates,
} from '../../widgets/query-rule-custom-data/query-rule-custom-data';
import Template from '../Template/Template';

type QueryRuleCustomDataProps = {
  cssClasses: QueryRuleCustomDataCSSClasses;
  templates: QueryRuleCustomDataTemplates;
  items: any[];
};

const QueryRuleCustomData = ({
  cssClasses,
  templates,
  items,
}: QueryRuleCustomDataProps) => (
  <Template
    templateKey="default"
    templates={templates}
    rootProps={{ className: cssClasses.root }}
    data={{ items }}
  />
);

export default QueryRuleCustomData;
