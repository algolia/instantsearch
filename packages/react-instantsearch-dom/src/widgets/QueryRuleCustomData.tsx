import React from 'react';
import { connectQueryRules } from 'react-instantsearch-core';

import PanelCallbackHandler from '../components/PanelCallbackHandler';
import QueryRuleCustomData from '../components/QueryRuleCustomData';

import type { QueryRuleCustomDataProps } from '../components/QueryRuleCustomData';

type CustomUserData = {
  [key: string]: any;
};

const QueryRuleCustomDataWidget: React.FC<
  QueryRuleCustomDataProps<CustomUserData>
> = (props) => (
  <PanelCallbackHandler {...props}>
    <QueryRuleCustomData {...props} />
  </PanelCallbackHandler>
);

export default connectQueryRules(QueryRuleCustomDataWidget, {
  $$widgetType: 'ais.queryRuleCustomData',
});
