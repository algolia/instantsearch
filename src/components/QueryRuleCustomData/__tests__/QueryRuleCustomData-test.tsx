/** @jsx h */

import { h } from 'preact';
import { render } from '@testing-library/preact';
import QueryRuleCustomData from '../QueryRuleCustomData';

type QueryRuleItem = {
  banner: string;
};

function defaultTemplate({ items }: { items: QueryRuleItem[] }) {
  return items.map(item => item.banner).join(' ');
}

describe('QueryRuleCustomData', () => {
  test('renders with empty items', () => {
    const items: QueryRuleItem[] = [];
    const props = {
      items,
      templates: {
        default: defaultTemplate,
      },
      cssClasses: {
        root: 'root',
      },
    };

    const { container } = render(<QueryRuleCustomData {...props} />);

    expect(container).toMatchSnapshot();
  });

  test('renders with items', () => {
    const items: QueryRuleItem[] = [
      { banner: 'image-1.png' },
      { banner: 'image-2.png' },
    ];
    const props = {
      items,
      templates: {
        default: defaultTemplate,
      },
      cssClasses: {
        root: 'root',
      },
    };

    const { container } = render(<QueryRuleCustomData {...props} />);

    expect(container).toMatchSnapshot();
  });
});
