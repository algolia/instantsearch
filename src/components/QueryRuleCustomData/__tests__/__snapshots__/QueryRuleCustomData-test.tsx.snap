// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryRuleCustomData renders with empty items 1`] = `
<Template
  data={Array []}
  rootProps={
    Object {
      "className": "root",
    }
  }
  rootTagName="div"
  templateKey="default"
  templates={
    Object {
      "default": "default",
    }
  }
  templatesConfig={Object {}}
  useCustomCompileOptions={Object {}}
/>
`;

exports[`QueryRuleCustomData renders with items 1`] = `
<Template
  data={
    Array [
      Object {
        "banner": "image-1.png",
      },
      Object {
        "banner": "image-2.png",
      },
    ]
  }
  rootProps={
    Object {
      "className": "root",
    }
  }
  rootTagName="div"
  templateKey="default"
  templates={
    Object {
      "default": "default",
    }
  }
  templatesConfig={Object {}}
  useCustomCompileOptions={Object {}}
/>
`;
